### PR TITLE
Need to apply hidden categories 

### LINKED ISSUE #4192: Hidden category not found by exact name
I apply a location-and-date category such as `[[Category:United Kingdom photographs taken on 2021-01-19]]` to all my images.

Presumably because these are hidden categories, they are not found by the app's category search.

In settings, please provide an option to have hidden categories found on the category search page (alternatively, please provide the option on that search page, so that it can be activated on an image-by-image basis).

If neither of these is possible please provide the option to enter the category name manually.

### PR DESCRIPTION
**Pull request by devarsh-mavani-19**
**Description (required)**
hidden categories should be shown when searched for example  `[[Category:United Kingdom photographs taken on 2021-01-19]]`
Fixes #4192

What changes did you make and why?

added API call that will fetch hidden category when entered by user.
Tested {build variant, e.g. ProdDebug} on {name of device or emulator} with API level {API level}.
tested on prodDebug
device: Redmi 6 pro
API 28
android 9 (pie)

### CODE REVIEW

**Comment by devarsh-mavani-19** [comment]
will add tests and proper javadoc as soon as I can.

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul can you please review this? thanks

**Comment by 4D17Y4** [comment]
When a PR is still under work, u can change the title to [WIP]<pr name> or make it a draft.

File: app/src/main/java/fr/free/nrw/commons/category/CategoriesModel.kt [hunk 1]
```diff
@@ -149,6 +150,13 @@ class CategoriesModel @Inject constructor(
         return categoryClient.searchCategories(title, SEARCH_CATS_LIMIT).toObservable()
     }
 
+    /**
+     * check if category exists
+     */
+    fun checkIfCategoryExists(query: String): Single<List<CategoryItem>> {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/category/CategoriesModel.kt:156]: Could you please rename to `checkWhetherCategoryExists`?
The name `checkIfCategoryExists` might be misunderstood as "_if category exists, then check_".

File: app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesPresenter.kt [hunk 2]
```diff
@@ -64,6 +78,14 @@ class CategoriesPresenter @Inject constructor(
             .subscribeOn(ioScheduler)
             .map { it.filterNot { categoryItem -> repository.containsYear(categoryItem.name) } }
 
+    /**
+     * checks if category exists
+     */
+    private fun checkCategoryExist(term: String) =
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesPresenter.kt:84]: How about `checkWhetherCategoryExists`?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Comment by nicolas-raoul** [comment]
With this PR, categories that would normally be suggested (suggestions that come from depictions for instance) appear only for a fraction of a second, before disappearing.

**Comment by nicolas-raoul** [comment]
1. Start uploading an image
2. At the depictions step, select depiction [Trégastel](https://www.wikidata.org/wiki/Q210421)
3. Go to the categories selection step
4. Problem: Normally the linked category should be suggested, but instead no categories are shown, the activity is blank
5. Type any letter, for instance "a", and the suggestions appear for a fraction of a second before disappearing.

**Comment by nicolas-raoul** [comment]
This seems to work with most depictions I tried.

At step 5, a better workaround is to just drag to the bottom (like if scrolling). That makes the suggestions appear and stay. I will now try to reproduce on master, to see whether it is really caused by this pull request or not.

**Comment by nicolas-raoul** [comment]
The problem seems to not exist on master.

**Comment by devarsh-mavani-19** [comment]
> 

hey, @nicolas-raoul I have fixed the issue. sorry for the delay.

**Comment by nicolas-raoul** [comment]
> means is it caused by this PR?

It seems to be the case indeed.
Are you able to reproduce the issue using [the steps above](https://github.com/commons-app/apps-android-commons/pull/4924#issuecomment-1080715139)?

**Comment by nicolas-raoul** [comment]
Sorry I had not seen your last comments, thank you for fixing!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Strangely I can't find https://commons.wikimedia.org/wiki/Category:United_Kingdom_photographs_taken_on_2021-01-19 by typing `United Kingdom photographs taken on 2021-01-19`. Am I missing something?
adb

**Comment by devarsh-mavani-19** [comment]
> Strangely I can't find https://commons.wikimedia.org/wiki/Category:United_Kingdom_photographs_taken_on_2021-01-19 by typing `United Kingdom photographs taken on 2021-01-19`. Am I missing something? adb

That's strange I tried and it worked for me
Screenshot_2022-03-30-09-15-51-111_fr free nrw commons

**Comment by devarsh-mavani-19** [comment]
@nicolas-raoul I tried to reproduce error on different device as well but it is not showing that error in my case

**Comment by nicolas-raoul** [comment]
Here is what logcat shows when I paste `United Kingdom photographs taken on 2021-01-19` in the categories search box:
```
04-01 23:47:57.629  5636  5803 I clp-JNI : Add pinfo.  (channel) 6a86011 PopupWindow:edae80d (server) (pkg) AppWindowToken{8d03d8c token=Token{14b72bf ActivityRecord{f9840de u0 fr.free.nrw.commons/.upload.UploadActivity t35203}}} 14585 10408 (action) 0
04-01 23:47:57.631 14585 14585 D ViewRootImpl@88cc510[PopupWindow:edae80d]: ViewPostIme pointer 0
04-01 23:47:57.685  5636  5803 I clp-JNI : Add pinfo.  (channel) 6a86011 PopupWindow:edae80d (server) (pkg) AppWindowToken{8d03d8c token=Token{14b72bf ActivityRecord{f9840de u0 fr.free.nrw.commons/.upload.UploadActivity t35203}}} 14585 10408 (action) 1
04-01 23:47:57.685 14585 14585 D ViewRootImpl@88cc510[PopupWindow:edae80d]: ViewPostIme pointer 1
04-01 23:47:57.689 14585 14585 D TextView: paste - clipdata type : HTML
04-01 23:47:57.700 14585 14585 D ViewRootImpl@88cc510[PopupWindow:edae80d]: dispatchDetachedFromWindow
04-01 23:47:57.704 14585 14585 D ViewRootImpl@88cc510[PopupWindow:edae80d]: Surface release. android.view.ViewRootImpl.doDie:7944 android.view.ViewRootImpl.die:7912 android.view.WindowManagerGlobal.removeViewLocked:497 android.view.WindowManagerGlobal.removeView:435 android.view.WindowManagerImpl.removeViewImmediate:124 android.widget.PopupWindow.dismissImmediate:2242 android.widget.PopupWindow.dismiss:2185 com.android.internal.widget.FloatingToolbar$FloatingToolbarPopup$4.onAnimationEnd:584 
04-01 23:47:57.708 14585 14585 D InputTransport: Input channel destroyed: fd=146
04-01 23:47:58.205 14585 15198 D OkHttp  : --> GET https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allcategories&prop=categoryinfo|description|pageimages&piprop=thumbnail&pithumbsize=70&gacprefix=United%20Kingdom%20photographs%20taken%20on%202021-01-19&gaclimit=25&gacoffset=0
04-01 23:47:58.206 14585 15198 D CookieManager: Domain:commons.wikimedia.org
04-01 23:47:58.660 14585 15198 D OkHttp  : <-- 200 https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&generator=allcategories&prop=categoryinfo|description|pageimages&piprop=thumbnail&pithumbsize=70&gacprefix=United%20Kingdom%20photographs%20taken%20on%202021-01-19&gaclimit=25&gacoffset=0 (455ms, 278-byte body)
```
then `No Categories found` appears.

**Comment by neslihanturan** [comment]
hi @devarsh-mavani-19 , thanks for the PR. Is this PR still active?

**Comment by nicolas-raoul** [comment]
Sorry for the delay @devarsh-mavani-19 would you mind rebasing (or pulling) from the latest main branch? Thanks!

**Comment by sivaraam** [comment]
Closing due to dormancy of the MR. Feel free to reopen if you're willing to work on this. 🙂

**Pull request closed without merging**